### PR TITLE
Fix for subscribe issue, on cold start always unsubscribed

### PR DESCRIPTION
**Pull request by mikechoch**
* During previous refactoring for USerDefault keys an issue was caused with SUBSCRIPTION key
* Two essential checks (OneSignal.m and OSSubscription.m) were changed and should not have been
* When pausing in app messaging, this only prevented items from being added to the queue, instead of from showing

* Demo app changes
  * Moved all UI property to ViewController.h interface so that the AppDelegate.m has access to them
  * Added a subscribe segment control (like consent and pausing IAM) to demo app, connected subscription observer
  * When subscription changes so does the segment control will also
  * Made segment control selected switches green instead
  * Fixed white text problem due to default in dark mode

* SDK changes
  * Reverted change for the SUBSCRIPTION key to fix the always unsubscribed issue
  * Using keyExists in OSSubscription.m now, replaces getSavedObjectForKey and objectForKey
  * Reverted changes for saving SUBSCRIPTION status in OneSignal.m
  * Moved the check for paused in app messaging so that queue items won't be shown now

---
This change is [Reviewable](https://reviewable.io/reviews/onesignal/onesignal-ios-sdk/599)

IMG_0007

### CODE REVIEW

File: iOS_SDK/OneSignalSDK/Source/OSMessagingController.m [hunk 1]
```diff
@@ -167,28 +159,31 @@ - (BOOL)isMessageInDisplayQueue:(NSString *)messageId {
...
-            // Show new IAM preview
             [self displayMessage:message];
         }
     };
 }
 
 - (void)displayMessage:(OSInAppMessage *)message {
+    // Check if the app disabled IAMs for this device before showing an IAM
+    if (_isInAppMessagingPaused) {
+        [OneSignal onesignal_Log:ONE_S_LL_VERBOSE message:@"In app messages will not show while pasued"];
```
Comment (jkasten2) [inline iOS_SDK/OneSignalSDK/Source/OSMessagingController.m]: Spelling issue, should be `paused`

File: iOS_SDK/OneSignalSDK/Source/OneSignalCommonDefines.h [hunk 2]
```diff
@@ -30,56 +30,70 @@
...
-#define DEPRECATED_SELECTORS @[@"application:didReceiveLocalNotification:", @"application:handleActionWithIdentifier:forLocalNotification:completionHandler:", @"application:handleActionWithIdentifier:forLocalNotification:withResponseInfo:completionHandler:"]
-#define USES_PROVISIONAL_AUTHORIZATION @"ONESIGNAL_USES_PROVISIONAL_PUSH_AUTHORIZATION"
-#define PERMISSION_HAS_PROMPTED @"OS_HAS_PROMPTED_FOR_NOTIFICATIONS_LAST"
-#define PERMISSION_ANSWERED_PROMPT @"OS_NOTIFICATION_PROMPT_ANSWERED_LAST"
-#define PERMISSION_ACCEPTED @"ONESIGNAL_ACCEPTED_NOTIFICATION_LAST"
-#define PERMISSION_PROVISIONAL_STATUS @"ONESIGNAL_PROVISIONAL_AUTHORIZATION_LAST"
-#define PERMISSION_PROVIDES_NOTIFICATION_SETTINGS @"OS_APP_PROVIDES_NOTIFICATION_SETTINGS"
-#define EXTERNAL_USER_ID @"OS_EXTERNAL_USER_ID"
+// OneSignalUserDefault keys
+// String values start with "OSUD_" to maintain a level of uniqueness fropm other apps
```
Comment (jkasten2) [inline iOS_SDK/OneSignalSDK/Source/OneSignalCommonDefines.h]: "uniqueness fropm other apps" should be "uniqueness from other libraries or app code"

File: iOS_SDK/OneSignalSDK/Source/OSPermission.m [hunk 3]
```diff
@@ -46,22 +46,22 @@ - (instancetype)initAsTo {
 
 - (instancetype)initAsFrom {
     let standardUserDefaults = OneSignalUserDefaults.initStandard;
-    _hasPrompted = [standardUserDefaults getSavedBoolForKey:PERMISSION_HAS_PROMPTED defaultValue:false];
-    _answeredPrompt = [standardUserDefaults getSavedBoolForKey:PERMISSION_ANSWERED_PROMPT defaultValue:false];
-    _accepted  = [standardUserDefaults getSavedBoolForKey:PERMISSION_ACCEPTED defaultValue:false];
-    _provisional = [standardUserDefaults getSavedBoolForKey:PERMISSION_PROVISIONAL_STATUS defaultValue:false];
-    _providesAppNotificationSettings = [standardUserDefaults getSavedBoolForKey:PERMISSION_PROVIDES_NOTIFICATION_SETTINGS defaultValue:false];
+    _hasPrompted = [standardUserDefaults getSavedBoolForKey:OSUD_WAS_PROMPTED_FOR_NOTIFICATIONS_LAST defaultValue:false];
+    _answeredPrompt = [standardUserDefaults getSavedBoolForKey:OSUD_PERMISSION_ANSWERED_PROMPT defaultValue:false];
```
Comment (jkasten2) [inline iOS_SDK/OneSignalSDK/Source/OSPermission.m]: Add `_LAST` to these 4 `initAsFrom` and `persistAsFrom` keys

File: iOS_SDK/OneSignalSDK/Source/OSPermission.m [hunk 4]
```diff
@@ -81,7 +81,7 @@ - (instancetype)copyWithZone:(NSZone*)zone {
 
 - (void)setHasPrompted:(BOOL)inHasPrompted {
     if (_hasPrompted != inHasPrompted)
-        [OneSignalUserDefaults.initStandard saveBoolForKey:HAS_PROMPTED_FOR_NOTIFICATIONS withValue:true];
+        [OneSignalUserDefaults.initStandard saveBoolForKey:OSUD_WAS_PROMPTED_FOR_NOTIFICATIONS withValue:true];
```
Comment (jkasten2) [inline iOS_SDK/OneSignalSDK/Source/OSPermission.m]: It seems `OSUD_WAS_PROMPTED_FOR_NOTIFICATIONS` is save here but don't see it read anywhere...

File: iOS_SDK/OneSignalSDK/Source/OSSubscription.m [hunk 5]
```diff
@@ -61,10 +61,10 @@ - (BOOL)compare:(OSSubscriptionState*)from {
 
 - (instancetype)initAsFrom {
     let standardUserDefaults = OneSignalUserDefaults.initStandard;
-    _accpeted = [standardUserDefaults getSavedBoolForKey:ACCEPTED_PERMISSION defaultValue:false];
-    _userId = [standardUserDefaults getSavedStringForKey:USERID_LAST defaultValue:nil];
-    _pushToken = [standardUserDefaults getSavedStringForKey:PUSH_TOKEN defaultValue:nil];
-    _userSubscriptionSetting = [standardUserDefaults getSavedObjectForKey:SUBSCRIPTION_SETTING defaultValue:nil] == nil;
+    _accpeted = [standardUserDefaults getSavedBoolForKey:OSUD_ACCEPTED_PERMISSION defaultValue:false];
```
Comment (jkasten2) [inline iOS_SDK/OneSignalSDK/Source/OSSubscription.m]: Should be ` OSUD_ACCEPTED_PERMISSION_LAST`

File: iOS_SDK/OneSignalSDK/Source/OneSignalCommonDefines.h (hunk 2 above)
Comment (mikechoch) [reply iOS_SDK/OneSignalSDK/Source/OneSignalCommonDefines.h]: Got it!

File: iOS_SDK/OneSignalSDK/Source/OSMessagingController.m (hunk 1 above)
Comment (mikechoch) [reply iOS_SDK/OneSignalSDK/Source/OSMessagingController.m]: Got it!

File: iOS_SDK/OneSignalSDK/Source/OSPermission.m (hunk 3 above)
Comment (mikechoch) [reply iOS_SDK/OneSignalSDK/Source/OSPermission.m]: Got it!

File: iOS_SDK/OneSignalSDK/Source/OSSubscription.m (hunk 5 above)
Comment (mikechoch) [reply iOS_SDK/OneSignalSDK/Source/OSSubscription.m]: Got it!

File: iOS_SDK/OneSignalSDK/Source/OSPermission.m (hunk 4 above)
Comment (mikechoch) [reply iOS_SDK/OneSignalSDK/Source/OSPermission.m]: Looking into this now, might need to go back in time for this one.

**Comment by mikechoch** [comment]
`testDisablingMessagesPreventsDisplay` is fixed now and passes. The issue was related to movement of the pause check for IAMs. We now still accumulate a message queue, but just wont show the IAMs in the app until IAM is enabled.

File: iOS_SDK/OneSignalSDK/Source/OSPermission.m (hunk 4 above)
Comment (mikechoch) [reply iOS_SDK/OneSignalSDK/Source/OSPermission.m]: Deleting this `saveBoolForKey`, but keeping the key for possible future usage

**Pull request merged**